rclpy/rclpy/parameter.py

```diff
@@ -211,3 +260,81 @@ def parameter_value_to_python(parameter_value: ParameterValue):
         raise RuntimeError(f'unexpected parameter type {parameter_value.type}')
 
     return value
+
+
+def parameter_dict_from_yaml_file(
+    parameter_file: str,
+    use_wildcard: bool = False,
+    target_nodes: Optional[List[str]] = None,
+    namespace: str = ''
+) -> Dict[str, ParameterMsg]:
+    """
+    Build a dict of parameters from a YAML file.
+
+    Will load all parameters if ``target_nodes`` is None or empty.
+
+    :raises RuntimeError: if a target node is not in the file
+    :raises RuntimeError: if the file is not in the same format as provided by ros2 param dump
+
+    :param parameter_file: Path to the YAML file to load parameters from.
+    :param use_wildcard: Use wildcard matching for the target nodes.
+    :param target_nodes: List of nodes in the YAML file to load parameters from.
+    :param namespace: Namespace to prepend to all parameters.
+    :return: A dict of Parameter messages keyed by the parameter names
+    """
+    with open(parameter_file, 'r') as f:
+        param_file = yaml.safe_load(f)
+        param_keys = []
+        param_dict = {}
+
+        if use_wildcard and '/**' in param_file:
+            param_keys.append('/**')
+
+        if target_nodes:
+            for n in target_nodes:
+                if n not in param_file.keys():
+                    raise RuntimeError(f'Param file does not contain parameters for {n},'
+                                       f'only for nodes: {list(param_file.keys())} ')
+                param_keys.append(n)
+        else:
+            # wildcard key must go to the front of param_keys so that
+            # node-namespaced parameters will override the wildcard parameters
+            keys = set(param_file.keys())
+            keys.discard('/**')
+            param_keys.extend(keys)
+
+        if len(param_keys) == 0:
+            raise RuntimeError('Param file does not contain selected parameters')
+
+        for n in param_keys:
+            value = param_file[n]
+            if type(value) != dict or 'ros__parameters' not in value:
+                raise RuntimeError('Invalid structure of parameter file for node {}'
+                                   'expected same format as provided by ros2 param dump'
```

Comment:
I missed this in earlier reviews, but I would change the error message here to avoid referencing "ros2 param dump".

Really, the error is that the yaml file is not following the expected format for a ROS parameters file.